File: R/xi.R
update_random_intercept <- function(prm, Y) {
  Ytilde <- Y - tcrossprod(rep(1, nrow(Y)), prm$alpha) -
    prm$Bt_eta - # time-varying main effect of chemicals latent factor
    prm$eta_int %*% prm$B # linear effect terms: interaction with coveriates, main effects of covariates
  
  update_random_intercept_cpp(prm$xi, Ytilde, prm$id, prm$uid,
                prm$Sigma, prm$sigmay_sqinv, prm$nu_sqinv,
                ncol(Y), length(prm$uid))
  # Update between subject variance
  # TODO: tune priors on a, b
  a <- 3.2 + length(prm$uid)/2
  b <- 1/50 + sum(apply(prm$xi, 1, function(x) t(x) %*% prm$Sigmainv %*% x))/2
  prm$nu_sqinv <- rgamma(1, shape=a, rate=b)
  stopifnot(prm$nu_sqinv > .Machine$double.eps)
  
  return(prm)
}

File: src/xi.cpp
#include <RcppArmadillo.h>
#include <math.h>
// [[Rcpp::depends(RcppArmadillo)]]
// [[Rcpp::plugins(cpp11)]]

using namespace std;
using namespace arma;

void update_random_intercept_cpp(arma::mat& xi, const arma::mat& Y,
                                 const arma::vec& id, const arma::vec& uid,
                                 const arma::mat& Sigma,
                                 const double sigmay_sqinv, const double nu_sqinv,
                                 int q, int n) {
  arma::mat Yi;
  arma::vec yi;
  for (int i = 0; i < n; i++) {
    Yi = Y.rows(find(id == uid(i)));
    yi = arma::sum(Yi, 0).t();
    int ti = Yi.n_rows;
    double si = 1/(nu_sqinv + ti * sigmay_sqinv);
    xi.row(i) = arma::mvnrnd(si * sigmay_sqinv * yi, si * Sigma).t();
  }
}
